lib/experimental/class-wp-html-updater.php

```diff
@@ -0,0 +1,556 @@
+			if ( ! $classes->has( $class_name ) ) {
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	public function set_attribute( $name, $value ) {
+		if ( $this->current_tag ) {
+			$escaped_new_value = esc_attr( $value );
+			$updated_attribute = "{$name}=\"{$escaped_new_value}\"";
+
+			$attr = $this->find_current_tag_attribute( $name );
+			if ( $attr ) {
+				// Update an existing attribute.
+				$diff = new WP_HTML_Diff(
+					$attr->get_start_index(),
+					$attr->get_end_index(),
+					$updated_attribute
+				);
+			} else {
+				// Create a new attribute at the tag's name end.
+				$diff = new WP_HTML_Diff(
+					$this->current_tag_name_ends_at,
+					$this->current_tag_name_ends_at,
+					" {$updated_attribute}"
+				);
+			}
+			$this->add_diff( $name, $diff );
+		}
+
+		return $this;
+	}
+
+	public function remove_attribute( $name ) {
+		if ( $this->current_tag ) {
+			$attr = $this->find_current_tag_attribute( $name );
+			if ( $attr ) {
+				$diff = new WP_HTML_Diff(
+					$attr->get_start_index(),
+					$attr->get_end_index(),
+					''
+				);
+				$this->add_diff( $name, $diff );
+			}
+		}
+
+		return $this;
+	}
+
+	private function find_current_tag_attribute( $name ) {
+		if ( array_key_exists( $name, $this->current_tag_attributes ) ) {
+			return $this->current_tag_attributes[ $name ];
+		}
+		do {
+			$attr = $this->parse_next_attribute();
+		} while ( $attr && ! $attr->name_equals( $name ) );
+
+		return $attr;
+	}
+
+	public function add_class( $class_name ) {
+		if ( $this->current_tag ) {
+			$this->current_class_names_set()->add( $class_name );
+		}
+
+		return $this;
+	}
+
+	public function remove_class( $class_name ) {
+		if ( $this->current_tag ) {
+			$this->current_class_names_set()->remove( $class_name );
+		}
+
+		return $this;
+	}
+
+	private function add_diff( $attribute_name, $diff ) {
+		if ( in_array( $attribute_name, $this->current_tag_touched_attributes_names, true ) ) {
+			throw new WP_HTML_Attribute_Already_Modified( "Only one change per tag attribute is supported and the attribute '{$attribute_name}' was already changed on tag {$this->current_tag}." );
+		}
+		$this->current_tag_touched_attributes_names[] = $attribute_name;
+		$this->diffs[]                                = $diff;
+	}
+
+	private function current_class_names_set() {
+		if ( $this->current_tag_class_names === null ) {
+			$class_attr                    = $this->find_current_tag_attribute( 'class' );
+			$class_value                   = $class_attr ? $class_attr->get_value() : '';
+			$this->current_tag_class_names = new WP_HTML_Updater_Classnames_Set( $class_value );
+		}
+
+		return $this->current_tag_class_names;
+	}
+
+	private function match( $regexp ) {
+		$matches = null;
+		$result  = preg_match(
+			$regexp,
+			$this->html,
+			$matches,
+			PREG_OFFSET_CAPTURE,
+			$this->parsed_characters
+		);
+		if ( 1 !== $result ) {
+			throw new WP_HTML_No_Match();
+		}
+
+		return $matches;
+	}
```

Comment:
as noted above we _might_ have significant performance overhead by creating this exception and a different return value could help avoid that.

further, _if_ we are going to abstract this here into a new function we _could_ take the additional step of trapping exceptions from the PCRE engine so they don't chain up the stack.

```php
// ignore PCRE exceptions such as PREG_INTERNAL_ERROR or PREG_BACKTRACK_LIMIT_ERROR
set_error_handler( function( $errno, $errmsg ) {} );
$result = preg_match( … );
restore_error_handler();
if ( 1 !== $result ) {
   return false;
}
```

in #42507 I have chosen to avoid trapping and ignoring those exceptions and also of abstracting the PCRE calls mainly to keep them concrete in the code for others to read, also because my IDE recognizes and syntax-highlights the PCRE patterns when found inside calls to the `preg_` functions.

Reply:
I also like the syntax highlighting part that this helper function takes away! Nonetheless, in this PR I chose to abstract this part to only reason about a single return value. In one of the recent commits, I also added a small convenience:

```php
$this->parsed_bytes = $this->index_after_match( $matches[0] );
```

As I noticed that each time I increment parsed_bytes by match length anyway.

I like the error trapping idea, too. I'm not sure, though, how likely are we to ever see these PCRE errors 🤔